### PR TITLE
[forge] local backend implementation

### PR DESCRIPTION
**Pull request by zihaoccc**
## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: testsuite/diem-swarm/src/swarm.rs [hunk 1]
```diff
@@ -121,6 +121,10 @@ impl DiemNode {
         self.peer_id
     }
 
+    pub fn stop(&mut self) {
+        self.process.kill().expect("node is not running");
+    }
```
Comment (bmwill) [inline testsuite/diem-swarm/src/swarm.rs:126]: So this probably won't be sufficient since then we won't ever be able to start back up the Node. We probably want to change the local Node to something like this:
```
#[derive(Debug)]
struct Process(Child);

impl Drop for Process {
    // When the Process struct goes out of scope we need to kill the child process
    fn drop(&mut self) {
        // check if the process has already been terminated
        match self.process.try_wait() {
            // The child process has already terminated, perhaps due to a crash
            Ok(Some(_)) => {}

            // The process is still running so we need to attempt to kill it
            _ => {
                self.process.kill().expect("Process wasn't running");
                self.process.wait().unwrap();
            }
        }
    }
}

struct LocalNode {
    diem_node_bin: PathBuf,
    process: Option<Process>,
    node_id: NodeId,
    peer_id: PeerId,
    log_path: PathBuf,
    config: NodeConfig,
}
```

Notice wrapping the process in an Option as a way to signal if the Node is running or not. It will let us start and stop the node while being able to reuse the Node struct.

File: testsuite/forge/src/backend/local.rs [hunk 2]
```diff
@@ -1,16 +1,128 @@
...
-    types::{chain_id::ChainId, AccountKey, LocalAccount},
+    types::{chain_id::ChainId, AccountKey, LocalAccount, PeerId},
 };
-use diem_swarm::swarm::DiemSwarm;
+use diem_swarm::swarm::{DiemNode, DiemSwarm, HealthStatus};
+
+struct LocalSwarm {
+    validator_swarm: DiemSwarm,
+    vfn_swarm: DiemSwarm,
+    pfn_swarm: DiemSwarm,
```
Comment (bmwill) [inline testsuite/forge/src/backend/local.rs:21]: Later I think we should look at fixing the underlying swarm type itself to be able to support the different types of nodes so we don't have to have multiple "swarms" like this. Course for right now if this works its a quick way to get something off the ground and we can always improve it later.

File: testsuite/forge/src/interface/public.rs [hunk 3]
```diff
@@ -111,16 +145,22 @@ impl Fund for Coffer<'_> {
...
+    tx_factory: TransactionFactory,
 }
 
 impl<'t> PublicInfo<'t> {
-    pub fn new(json_rpc_url: &'t str, chain_id: ChainId, coffer: Coffer<'t>) -> Self {
+    pub fn new(
+        json_rpc_url: &'t str,
+        chain_id: ChainId,
+        coffer: Coffer<'t>,
+        tx_factory: TransactionFactory,
```
Comment (bmwill) [inline testsuite/forge/src/interface/public.rs:157]: I wouldn't require a tx factory to build this type. instead we should build it internally since all you need is the chain_id which is passed in here.

File: testsuite/forge/src/interface/public.rs [hunk 4]
```diff
@@ -44,19 +45,41 @@ impl<'t> PublicUsageContext<'t> {
...
         &mut self,
         currency: Currency,
         auth_key: AuthenticationKey,
         amount: u64,
+        tx_factory: TransactionFactory,
+    ) -> Result<()> {
+        self.public_info
+            .coffer
+            .fund(currency, auth_key, amount, tx_factory)
+    }
```
Comment (bmwill) [inline testsuite/forge/src/interface/public.rs:62]: I don't think the fund function should be changed, there's no real reason for it and it actually ends up leaking the abstraction. specifically the "funding" can be done either through a faucet service or via the Coffer actually having the mint key but the caller shouldn't care about this implementation detail, so i would change back the fund function, trait and coffer.

File: testsuite/forge/src/interface/public.rs [hunk 5]
```diff
@@ -44,19 +45,41 @@ impl<'t> PublicUsageContext<'t> {
         self.public_info.chain_id
     }
 
+    pub fn tx_factory(&self) -> TransactionFactory {
+        self.public_info.tx_factory.clone()
+    }
```
Comment (bmwill) [inline testsuite/forge/src/interface/public.rs:50]: instead of cloning return a `&TransactionFactory` and let the caller determine if they want to clone it

File: testsuite/forge/src/main.rs [hunk 6]
```diff
@@ -72,6 +77,79 @@ impl PublicUsageTest for FundAccount {
...
+        client.batch(batch)?;
+        client.wait_for_signed_transaction(&create_payer, None, None)?;
+        client.wait_for_signed_transaction(&create_payee, None, None)?;
+        let account_view = client.get_account(payer.address())?.into_inner().unwrap();
+        let balance = account_view
+            .balances
+            .iter()
+            .find(|b| b.currency == currency)
+            .unwrap();
+        assert_eq!(balance.amount, 100);
```
Comment (bmwill) [inline testsuite/forge/src/main.rs:131]: Maybe we can add a helper function to do this balance checking, since the code is repeated three times.

**Review by bmwill — COMMENTED** [review]
Really great start! I think the ultimate test of these APIs will be when we implement a Network test and actually try to do something interesting with the running nodes.

File: testsuite/forge/src/main.rs (hunk 6 above)
Comment (zihaoccc) [reply testsuite/forge/src/main.rs:131]: yea, my later change has some checking fns includes balance etc, it will be in next commit

File: testsuite/diem-swarm/src/swarm.rs (hunk 1 above)
Comment (zihaoccc) [reply testsuite/diem-swarm/src/swarm.rs:126]: this is a brilliant way to determine node status

File: testsuite/forge/src/backend/local.rs (hunk 2 above)
Comment (zihaoccc) [reply testsuite/forge/src/backend/local.rs:21]: created track ticket https://github.com/diem/diem/issues/8269 for afterward improvement

File: testsuite/forge/src/interface/public.rs (hunk 4 above)
Comment (zihaoccc) [reply testsuite/forge/src/interface/public.rs:62]: oh, this should be the display format here, all the change for fund here is one more tx_factory pass in, it will be removed since addressed other comments

Comment (zihaoccc) [reply testsuite/forge/src/interface/public.rs:62]: My thought is I do not really prefer let Coffer to hold a tx_factory, I more like to make tx_factory in Context, so I removed tx_factory out of Coffer::TreasuryCompliance

File: testsuite/forge/src/interface/public.rs [hunk 7]
```diff
@@ -65,24 +87,35 @@ pub enum Coffer<'t> {
...
 
 pub trait Fund {
-    fn fund(&mut self, currency: Currency, auth_key: AuthenticationKey, amount: u64) -> Result<()>;
+    fn fund(
+        &mut self,
+        currency: Currency,
+        auth_key: AuthenticationKey,
+        amount: u64,
+        tx_factory: TransactionFactory,
+    ) -> Result<()>;
```
Comment (bmwill) [inline testsuite/forge/src/interface/public.rs:96]: Mind if we just revert these changes to the Fund trait for now. It doesn't really seem necessary atm

File: testsuite/forge/src/interface/node.rs [hunk 8]
```diff
@@ -49,10 +61,7 @@ pub trait Node {
     fn clear_storage(&mut self) -> Result<()>;
 
     /// Performs a Health Check on the Node
-    fn health_check(&self) -> Result<()>;
-
-    /// Performs a connectivity check on this Node
-    fn connectivity_check(&self) -> Result<()>; // Maybe have this as a function as a part of a validator or fullnode
```
Comment (bmwill) [inline testsuite/forge/src/interface/node.rs:55]: What's the rationale for getting rid of the connectivity check?

File: testsuite/forge/src/interface/node.rs [hunk 9]
```diff
@@ -49,10 +61,7 @@ pub trait Node {
...
+    fn health_check(&mut self) -> HealthStatus;
```
Comment (bmwill) [inline testsuite/forge/src/interface/node.rs:64]: This makes sense, although we probably want to change the health_check function for the local node to never panic

**Review by bmwill — COMMENTED** [review]
Just a couple of nits then we can get this landed

File: testsuite/forge/src/interface/node.rs (hunk 8 above)
Comment (zihaoccc) [reply testsuite/forge/src/interface/node.rs:55]: since we are current using diem-node check_connectivity, I removed this place holder at this stage, we will add this fn back after we solved the diem-swarm underlying for better place

**Review by bmwill — APPROVED** [review]
Looks like you'll just need to fix some lint errors
